### PR TITLE
cephadm: also print occupying process of port in use

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -985,6 +985,34 @@ def check_ip_port(ctx, ip, port):
...
 
+def _parse_ss_lpnt(data:str):
+    # ss doen't support json output. Thus we have to parse it.
+    # https://unix.stackexchange.com/questions/157823/list-ports-a-process-pid-is-listening-on-preferably-using-iproute2-tools
+    out = []
+    for line in data.splitlines():
+        if line.startswith('STATE'):
+            if ' '.join(line.split()) != 'State Recv-Q Send-Q Local Address:Port Peer Address:Port':
+                logger.warning('ss -lpnt: unable to parse header "{line}"')
+                return {}
```
Comment (mgfritch) [inline src/cephadm/cephadm:996]: use the `-H, -no-header` option?

**Review by mgfritch — COMMENTED** [review]
this adds yet another dependency for cephadm ... add logic to `check-host` and `prepare-host` for the additional `ss` binary?

**Comment by sebastian-philipp** [comment]
> this adds yet another dependency for cephadm ... add logic to `check-host` and `prepare-host` for the additional `ss` binary?

I'm just using it for showing a better error message. I'd rather ignore the dependency and fail gracefully.

File: src/cephadm/cephadm (hunk 1 above)
Comment (sebastian-philipp) [reply src/cephadm/cephadm:996]: how do I then verify that the 4th column is the local address port?

**Review by p-se — APPROVED** [review]
lgtm

I like it very much, as it seems to be a soft dependency only but increases the usability!

File: src/cephadm/cephadm [hunk 2]
```diff
@@ -985,6 +985,34 @@ def check_ip_port(ctx, ip, port):
...
+            continue
+        before, after = line.split('users:((')
+        data = dict(zip('State,Recv-Q,Send-Q,Local Address:Port,Peer Address:Port'.split(','), before.split()))
+
+        users = after.rstrip('))').split(',')
+        data.update(dict(zip('name pid'.split(), users[:2])))
+        out.append(data)
+    return out
+
+def port_in_use_by(ctx, port_num: int) -> str:
```
Comment (jmolmo) [inline src/cephadm/cephadm:1007]: I suggest to use re.
ex:
```python
def port_used_by( port_num: int):
    out, err, code = call(ctx, ['ss', '-l', '-p', '-n', '-t'])
    pattern = '\d+:' + port_num + '\s+\d+.\d+.\d+.\d+.+users:\(\(\"(?P<process>.*)\",'
    used = re.search(pattern, out)
    if used:
        return used.group('process')
     else:
         return  ""
```

File: src/cephadm/tests/test_cephadm.py [hunk 3]
```diff
@@ -305,6 +305,138 @@ def test_last_local_images(self):
         image = cd._filter_last_local_ceph_image(out)
         assert image == 'docker.io/ceph/ceph:v15.2.5'
 
+    def test_parse_ss_lpnt(self):
```
Comment (jmolmo) [inline src/cephadm/tests/test_cephadm.py:308]: I think the the test can be centered in if the port is in the "ss" output,  or not.

**Review by jmolmo — CHANGES_REQUESTED** [review]
Probably using regular expressions the code can be simplified. 
The test is just checking the split of lines... i think we must check if the port is or not in the ss output

File: src/cephadm/cephadm (hunk 2 above)
Comment (sebastian-philipp) [reply src/cephadm/cephadm:1007]: building something that also works for IPv6 is going to be painful. and in the end I'd opt for a reg ex that simply splits by `\s`

**Comment by sebastian-philipp** [comment]
replaced by #39421

**Pull request closed without merging**